### PR TITLE
refactor(frontend): rewrite 3 system functions using context

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR rewrites 3 frontend functions using `#[function]` with context:

- `pg_relation_size`
- `pg_indexes_size`
- `pg_get_userbyid`

Currently these functions are implemented by rewritten to subqueries on system table, which is verbose and hard to maintain.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

### CODE REVIEW

**Comment by wangrunji0408** [comment]
```
thread 'rw-main' has overflowed its stack
fatal runtime error: stack overflow
```

The stack overflow (on debug mode) is caused by the following code:

https://github.com/risingwavelabs/risingwave/blob/144286099fc189f70eb716159fc2bd7edc10c1d5/src/frontend/src/scheduler/local.rs#L180-L185

If you print the size of each `exec`:

```
2848
5728
11520
23200
46432
92928
```

It seems the size will double each time it is nested. 🤯

cc @TennyZhuang @shanicky @yezizp2012 @st1page

**Review by xxchan — COMMENTED** [review]
This is so cool! Context is quite useful for system functions.

**Comment by wangrunji0408** [comment]
> QQ: I remember we force queries scanning system tables running in local mode. Do we do so for functions with context? (Do we need to?)

Yes. I guess we have not forbidden system functions in distributed mode or in mviews. But doing so will result in an `context not found` error or null output.

File: src/frontend/src/scheduler/local.rs [hunk 1]
```diff
@@ -173,14 +174,16 @@ impl LocalQueryExecution {
...
-        let exec = async move { SEARCH_PATH::scope(search_path, exec).await };
-        let exec = async move { AUTH_CONTEXT::scope(auth_context, exec).await };
-        let exec = async move { TIME_ZONE::scope(time_zone, exec).await };
+        // box is necessary, otherwise the size of `exec` will double each time it is nested.
+        let exec = async move { CATALOG_READER::scope(catalog_reader, exec).await }.boxed();
+        let exec = async move { USER_INFO_READER::scope(user_info_reader, exec).await }.boxed();
+        let exec = async move { DB_NAME::scope(db_name, exec).await }.boxed();
+        let exec = async move { SEARCH_PATH::scope(search_path, exec).await }.boxed();
+        let exec = async move { AUTH_CONTEXT::scope(auth_context, exec).await }.boxed();
+        let exec = async move { TIME_ZONE::scope(time_zone, exec).await }.boxed();
```
Comment (xxchan) [inline src/frontend/src/scheduler/local.rs:186]: What about using a HashMap or Tuple as the `LocalKey`? 🤣 (TBH I don't know which way is better)

**Comment by BugenZhao** [comment]
> It seems the size will double each time it is nested. 🤯

We've encountered this issue before! 😄

https://github.com/risingwavelabs/risingwave/blob/1c52ee772c97218257e50709f8fdf3ccd34d9962/src/utils/pgwire/src/pg_protocol.rs#L244-L246

This blog could be a good reference: https://swatinem.de/blog/future-size/

File: src/frontend/src/scheduler/local.rs (hunk 1 above)
Comment (wangrunji0408) [reply src/frontend/src/scheduler/local.rs:186]: I don't know either. I think collecting them to [`std::error::Request`](https://doc.rust-lang.org/std/error/struct.Request.html) may be a more elegant solution.

**Review by xxchan — APPROVED** [review]
Generally LGTM. Disclaimer: I haven't reviewed whether the functions' logic is correct.

**Review by TennyZhuang — APPROVED** [review]

File: src/frontend/src/scheduler/local.rs [hunk 2]
```diff
@@ -177,12 +177,13 @@ impl LocalQueryExecution {
             AUTH_CONTEXT, CATALOG_READER, DB_NAME, SEARCH_PATH, USER_INFO_READER,
         };
 
-        let exec = async move { CATALOG_READER::scope(catalog_reader, exec).await };
-        let exec = async move { USER_INFO_READER::scope(user_info_reader, exec).await };
-        let exec = async move { DB_NAME::scope(db_name, exec).await };
-        let exec = async move { SEARCH_PATH::scope(search_path, exec).await };
-        let exec = async move { AUTH_CONTEXT::scope(auth_context, exec).await };
-        let exec = async move { TIME_ZONE::scope(time_zone, exec).await };
+        // box is necessary, otherwise the size of `exec` will double each time it is nested.
```
Comment (BugenZhao) [inline src/frontend/src/scheduler/local.rs:180]: I suppose boxing `exec` is enough, making it possible to inline these scoped futures.

**Pull request merged**